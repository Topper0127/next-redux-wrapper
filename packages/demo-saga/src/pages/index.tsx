import React from 'react';
import {useSelector} from 'react-redux';
import {NextPage} from 'next';
import {State} from '../components/reducer';
import {SAGA_ACTION} from '../components/saga';

export interface ConnectedPageProps {
    custom: string;
}

// Page itself is not connected to Redux Store, it has to render Provider to allow child components to connect to Redux Store
const Page: NextPage<ConnectedPageProps> = ({custom}: ConnectedPageProps) => {
    const {page} = useSelector<State, State>(state => state);
    return (
        <div className="index">
            <pre>{JSON.stringify({page, custom}, null, 2)}</pre>
        </div>
    );
};

Page.getInitialProps = ({store}) => {
    store.dispatch({type: SAGA_ACTION});
    return {custom: 'custom'};
};

export default Page;
