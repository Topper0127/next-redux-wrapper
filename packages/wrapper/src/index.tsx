import React, {Component} from 'react';
import {Store} from 'redux';
import {NextComponentType, NextContext} from 'next';
import {NextAppContext} from 'next/app';

const _debug = false;
const DEFAULT_KEY = '__NEXT_REDUX_STORE__';
const isServer = typeof window === 'undefined';

const initStore = ({makeStore, initialState, config, ctx}: InitStoreOptions): Store => {
    const {storeKey} = config;

    const createStore = () =>
        makeStore(config.deserializeState(initialState), {
            ...ctx,
            ...config,
            isServer
        });

    if (isServer) return createStore();

    // Memoize store if client
    if (!window[storeKey]) {
        window[storeKey] = createStore();
    }

    return window[storeKey];
};

export default (makeStore: MakeStore, config?: Config) => {
    config = {
        storeKey: DEFAULT_KEY,
        debug: _debug,
        serializeState: state => state,
        deserializeState: state => state,
        ...config
    };

    return (App: NextComponentType | any) =>
        class WrappedApp extends Component<WrappedAppProps> {
            static displayName = `withRedux(${App.displayName || App.name || 'App'})`;

            static getInitialProps = async (appCtx: NextJSAppContext) => {
                if (!appCtx) throw new Error('No app context');
                if (!appCtx.ctx) throw new Error('No page context');

                const store = initStore({
                    makeStore,
                    config,
                    ctx: appCtx.ctx
                });

                if (config.debug)
                    console.log('1. WrappedApp.getInitialProps wrapper got the store with state', store.getState());

                appCtx.ctx.store = store;
                appCtx.ctx.isServer = isServer;

                let initialProps = {};

                if ('getInitialProps' in App) {
                    initialProps = await App.getInitialProps.call(App, appCtx);
                }

                if (config.debug) console.log('3. WrappedApp.getInitialProps has store state', store.getState());

                return {
                    isServer,
                    initialState: config.serializeState(store.getState()),
                    initialProps
                };
            };

            constructor(props, context) {
                super(props, context);

                const {initialState} = props;

                if (config.debug) console.log('4. WrappedApp.render created new store with initialState', initialState);

                this.store = initStore({
                    makeStore,
                    initialState,
                    config
                });
            }

            store: Store;

            render() {
                const {initialProps, initialState, ...props} = this.props;

                // Cmp render must return something like <Provider><Component/></Provider>
                return <App {...props} {...initialProps} store={this.store} />;
            }
        };
};

export interface Config {
    serializeState?: (any) => any;
    deserializeState?: (any) => any;
    storeKey?: string;
    debug?: boolean;
}

export interface NextJSContext extends NextContext {
    store: Store;
    isServer: boolean;
}

export interface NextJSAppContext extends NextAppContext {
    ctx: NextJSContext;
}

export interface MakeStoreOptions extends Config, NextJSContext {
    isServer: boolean;
}

export declare type MakeStore = (initialState: any, options: MakeStoreOptions) => Store;

export interface InitStoreOptions {
    makeStore: MakeStore;
    config: Config;
    initialState?: any;
    ctx?: NextJSContext;
}

export interface WrappedAppProps {
    initialProps: any; // stuff returned from getInitialProps
    initialState: any; // stuff in the Store state after getInitialProps
    isServer: boolean;
}

export interface AppProps {
    store: Store;
}
