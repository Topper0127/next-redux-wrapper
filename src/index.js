import React, {Component} from "react";

let _debug = false;
const DEFAULT_KEY = '__NEXT_REDUX_STORE__';
const isServer = typeof window === 'undefined';

/**
 * @param makeStore
 * @param initialState
 * @param config
 * @param ctx
 * @return {{getState: function, dispatch: function}}
 */
const initStore = ({makeStore, initialState, config, ctx = {}}) => {

    const {storeKey} = config;

    const createStore = () => makeStore(
        config.deserializeState(initialState),
        {
            ...ctx,
            ...config,
            isServer
        }
    );

    if (isServer) return createStore();

    // Memoize store if client
    if (!window[storeKey]) {
        window[storeKey] = createStore();
    }

    return window[storeKey];

};

/**
 * @param makeStore
 * @param config
 * @return {function(App): {getInitialProps, new(): WrappedApp, prototype: WrappedApp}}
 */
export default (makeStore, config = {}) => {

    config = {
        storeKey: DEFAULT_KEY,
        debug: _debug,
        serializeState: state => state,
        deserializeState: state => state,
        ...config
    };

    return (App) => (class WrappedApp extends Component {

        static displayName = `withRedux(${App.displayName || App.name || 'App'})`;

        static getInitialProps = async (appCtx) => {

            if (!appCtx) throw new Error('No app context');
            if (!appCtx.ctx) throw new Error('No page context');

            const store = initStore({
                makeStore,
                config,
                ctx: appCtx.ctx
            });

            if (config.debug) console.log('1. WrappedApp.getInitialProps wrapper got the store with state', store.getState());

            appCtx.ctx.store = store;
            appCtx.ctx.isServer = isServer;

            let initialProps = {};

            if ('getInitialProps' in App) {
                initialProps = await App.getInitialProps.call(App, appCtx);
            }

            if (config.debug) console.log('3. WrappedApp.getInitialProps has store state', store.getState());

            return {
                isServer,
                initialState: config.serializeState(store.getState()),
                initialProps: initialProps
            };

        };

        constructor(props, context) {

            super(props, context);

            const {initialState} = props;

            if (config.debug) console.log('4. WrappedApp.render created new store with initialState', initialState);

            this.store = initStore({
                makeStore,
                initialState,
                config
            });

        }

        render() {

            let {initialProps, initialState, ...props} = this.props;

            // Cmp render must return something like <Provider><Component/></Provider>
            return (
                <App {...props} {...initialProps} store={this.store}/>
            );

        }

    });

};