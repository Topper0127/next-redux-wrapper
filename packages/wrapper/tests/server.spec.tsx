import * as React from 'react';
import {create} from 'react-test-renderer';
import {createWrapper} from '../src';
import {DummyComponent, makeStore, makeStoreStub} from './testlib';
import {NextPageContext} from 'next';

describe('function API', () => {
    test.each([
        ['getInitialAppProps', 'simple context', 1, {pageProps: {prop: 'val'}}, {ctx: {req: 'request'}}],
        ['getInitialPageProps', 'simple context', 1, {pageProps: {prop: 'val'}}, {req: 'request'}],
        ['getInitialPageProps', 'context with store', 0, {pageProps: {prop: 'val'}}, {req: 'request', store: 'foo'}],
        [
            'getStaticProps',
            'simple context',
            1,
            {props: {prop: 'val'}, revalidate: true},
            {preview: true, previewData: 'previewData'},
        ],
        ['getServerSideProps', 'simple context', 1, {props: {prop: 'val'}}, {req: 'request'}],
    ])('%s (%s)', async (func: string, subtype: string, calledTime: number, props: any, context: any) => {
        const makeStore = makeStoreStub({state: 'val'});
        const wrapper = createWrapper(makeStore);

        const propsCb = jest.fn(async context => {
            expect(makeStore).toBeCalledTimes(calledTime);
            expect(context).toMatchSnapshot();
            return props;
        });

        // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
        //@ts-ignore
        const fn = wrapper[func](propsCb);
        expect(await fn(context)).toMatchSnapshot();
        expect(propsCb).toBeCalledTimes(1);
    });

    test('getInitialPageProps context with store & no callback', async () => {
        const makeStore = makeStoreStub({state: 'val'});
        const wrapper = createWrapper(makeStore);
        expect(
            await wrapper.getInitialPageProps()({
                // purposely make a store to enter condition
                store: makeStore(),
            } as any),
        ).toMatchSnapshot();
        expect(makeStore).toBeCalledTimes(1); // make sure makeStore has not been called in wrapper
    });
});

describe('withRedux', () => {
    describe('merges props', () => {
        test('for page case', () => {
            const wrapper = createWrapper(makeStore);
            const WrappedPage: any = wrapper.withRedux(DummyComponent);
            expect(
                create(<WrappedPage initialProps={{fromPage: true}} somePropFromNextJs={true} />).toJSON(),
            ).toMatchSnapshot();
        });
        test('for app case', () => {
            const wrapper = createWrapper(makeStore);
            const WrappedApp: any = wrapper.withRedux(DummyComponent);
            expect(
                create(
                    <WrappedApp initialProps={{pageProps: {fromApp: true}}} pageProps={{getStaticProp: true}} />,
                ).toJSON(),
            ).toMatchSnapshot();
        });
    });
    describe('wrapped component', () => {
        test('should not have getInitialProps if source component did not have it', () => {
            const wrapper = createWrapper(makeStore);
            expect(wrapper.withRedux(DummyComponent)).not.toHaveProperty('getInitialProps');
        });
        test('should have getInitialProps if source component have it', async () => {
            const wrapper = createWrapper(makeStore);

            const Page = (props: any) => <DummyComponent {...props} />;
            Page.getInitialProps = () => ({prop: 'val'});
            const WrappedPage = wrapper.withRedux(Page) as any;
            expect(await WrappedPage.getInitialProps()).toMatchSnapshot();
            expect(create(<WrappedPage />).toJSON()).toMatchSnapshot(); // to trigger creation of class
        });
    });
});

describe('custom serialization', () => {
    test('serialize on server and deserialize on client', async () => {
        const wrapper = createWrapper(makeStore, {
            serializeState: (state: any) => ({...state, serialized: true}),
            deserializeState: (state: any) => ({...state, deserialized: true}),
            debug: true,
        });

        const props = await wrapper.getInitialPageProps()({} as NextPageContext);

        expect(props).toMatchSnapshot();

        const WrappedApp = wrapper.withRedux(DummyComponent);

        expect(create(<WrappedApp {...props} />).toJSON()).toMatchSnapshot();
    });
});
