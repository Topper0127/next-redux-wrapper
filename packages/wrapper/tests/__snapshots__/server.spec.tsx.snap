// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`custom serialization serialize on server and deserialize on client 1`] = `
Object {
  "initialProps": Object {},
  "initialState": Object {
    "reduxStatus": "init",
    "serialized": true,
  },
}
`;

exports[`custom serialization serialize on server and deserialize on client 2`] = `
<div>
  {"props":{},"state":{"reduxStatus":"init","serialized":true,"deserialized":true}}
</div>
`;

exports[`function API getInitialAppProps (simple context) 1`] = `
Object {
  "ctx": Object {
    "req": "request",
    "store": Object {
      "dispatch": [MockFunction],
      "getState": [MockFunction],
      "subscribe": [MockFunction],
    },
  },
}
`;

exports[`function API getInitialAppProps (simple context) 2`] = `
Object {
  "initialProps": Object {
    "pageProps": Object {
      "prop": "val",
    },
  },
  "initialState": Object {
    "state": "val",
  },
}
`;

exports[`function API getInitialPageProps (context with store) 1`] = `
Object {
  "req": "request",
  "store": "foo",
}
`;

exports[`function API getInitialPageProps (context with store) 2`] = `
Object {
  "pageProps": Object {
    "prop": "val",
  },
}
`;

exports[`function API getInitialPageProps (simple context) 1`] = `
Object {
  "req": "request",
  "store": Object {
    "dispatch": [MockFunction],
    "getState": [MockFunction],
    "subscribe": [MockFunction],
  },
}
`;

exports[`function API getInitialPageProps (simple context) 2`] = `
Object {
  "initialProps": Object {
    "pageProps": Object {
      "prop": "val",
    },
  },
  "initialState": Object {
    "state": "val",
  },
}
`;

exports[`function API getInitialPageProps context with store & no callback 1`] = `undefined`;

exports[`function API getServerSideProps (simple context) 1`] = `
Object {
  "req": "request",
  "store": Object {
    "dispatch": [MockFunction],
    "getState": [MockFunction],
    "subscribe": [MockFunction],
  },
}
`;

exports[`function API getServerSideProps (simple context) 2`] = `
Object {
  "props": Object {
    "initialState": Object {
      "state": "val",
    },
    "prop": "val",
  },
}
`;

exports[`function API getStaticProps (simple context) 1`] = `
Object {
  "preview": true,
  "previewData": "previewData",
  "store": Object {
    "dispatch": [MockFunction],
    "getState": [MockFunction],
    "subscribe": [MockFunction],
  },
}
`;

exports[`function API getStaticProps (simple context) 2`] = `
Object {
  "props": Object {
    "initialState": Object {
      "state": "val",
    },
    "prop": "val",
  },
  "revalidate": true,
}
`;

exports[`withRedux merges props for app case 1`] = `
<div>
  {"props":{"pageProps":{"fromApp":true,"getStaticProp":true}},"state":{"reduxStatus":"init"}}
</div>
`;

exports[`withRedux merges props for page case 1`] = `
<div>
  {"props":{"fromPage":true,"somePropFromNextJs":true},"state":{"reduxStatus":"init"}}
</div>
`;

exports[`withRedux wrapped component should have getInitialProps if source component have it 1`] = `
Object {
  "prop": "val",
}
`;

exports[`withRedux wrapped component should have getInitialProps if source component have it 2`] = `
<div>
  {"props":{},"state":{"reduxStatus":"init"}}
</div>
`;
